xmas-lights-flask/app.py

```diff
@@ -0,0 +1,225 @@
+                except Exception as exception:
+                    reset = True
+                    error_string = traceback.format_exc(limit=3)
+                    current_pattern['error'] = error_string
+                    current_pattern['active'] = False
+                else:
+                    if led_index == 0:
+                        # led.on() if result[1] > 127 else led.off()
+                        pass
+                    # led_strip.set_pixel(led_index, result)  # TODO
+                if reset:
+                    break
+
+            if not reset:
+                pass
+                # led_strip.show()  # TODO
+
+app = Flask(__name__)
+
+@app.route("/",methods=['GET'])
+def home():
+    global patterns
+    return app_html.generate_html(json.dumps(json.dumps(patterns)))
+    
+@app.route("/", methods=['POST', 'DELETE'])
+def update_pattern():
+        global reset
+        global patterns
+
+        if request.method == 'DELETE':
+            try:
+                del patterns[post_data['id']]
+            except KeyError:
+                return ("Invalid Pattern ID", 400)
+            else:
+                with open('patterns.json', 'w') as file:
+                    file.write(json.dumps(patterns))
+            return "OK"
+
+        elif request.method == 'POST':
+            json_keys = {
+                'active': bool,
+                'name': str,
+                'author': str,
+                'script': str,
+            }
+
+            post_data = request.json
+
+            def key_valid(post_data, key, key_type):
+                return key in post_data and isinstance(post_data[key], key_type)
+
+            if key_valid(post_data, 'id', str):
+                pattern_id = request.json['id']
+            else:
+                pattern_id = str(random.getrandbits(32))
+
+            # allow partial update if pattern_id already exists
+            if pattern_id in patterns or all(key_valid(post_data, key, key_type) for key, key_type in json_keys.items()):
+                if pattern_id not in patterns:
+                    patterns[pattern_id] = {}
+                do_reset = patterns[pattern_id].get('active', False) or post_data.get('active', False)
+                for key, key_type in json_keys.items():
+                    if key_valid(post_data, key, key_type):
+                        if key == 'active' and post_data[key] is True:
+                            patterns[pattern_id]['error'] = None
+                            for pattern in patterns.values():
+                                pattern['active'] = False
+                        patterns[pattern_id][key] = post_data[key]
+
+                if do_reset:
+                    reset = True
+
+                with open('patterns.json', 'w') as file:
+                    file.write(json.dumps(patterns))
+
+            return "OK"
+    
+
+if __name__=="__main__":
+    threading.Thread(target=led_thread).start()
```

Comment:
Didn't have much joy with concurrent.futures.ThreadedPoolExecutor instead here, so stuck to simpler API